### ISSUE TITLE
Is it worth writing up an RFC on adding new escape characters?

### ISSUE THREAD

**Issue by rkeithhill** [PowerShell/PowerShell#2751]
Given the inherent breaking changes between Windows PowerShell 5.x and PowerShell Core, especially when running on Linux and macOS, can we consider adding a few new escape characters?

With the addition of ANSI esc sequences to the Windows console host, and existence of the same on terminals in Linux and macOS,  I would like to have the following esc chars added:
 ```
`e - ASCII escape char 0x1b
`u - specify Unicode character `u<hex-digit>+ e.g. `u2261
`_ - more "visible/obvious" line continuation
```
OK, so that last one is a stretch but there seems to be a few problems with line continuation in PowerShell - hard to see / trailing whitespace.  Maybe PowerShell Core presents an opportunity to tweak line continuation a little??

If the PowerShell Committee is willing to entertain this idea, is it worth writing up as an RFC?

**Comment by lzybkr** [comment]
I suppose it might break some script obfuscation techniques:

```powershell
`G`et`-`C`h`i`l`d`It`e`m
```

But it does seem like a valuable addition.  We don't need the RFC just yet, but I tagged this for committee review.

**Comment by joeyaiello** [comment]
I'm also okay with ``` `e``` and ``` `u``` without much discussion. Not sure about ``` `_```, but we'll discuss it in our next committee meeting.

**Comment by rkeithhill** [comment]
> Not sure about `_

Me neither.  Just throwing it out because I do think line continuation in PowerShell is one of its weaker areas.  At one time, I was suggesting just using `_` until I realized that duh, `_` is a valid function name. :-)

**Comment by powercode** [comment]
Maybe the line continuation problem is worthy of an issue of it's own?

**Comment by chuanjiao10** [comment]
This is the python2 misconception: in the same script into a variety of coding, it is easy to garbled.

BOM header + unicode,BOM header + utf8,is good enough

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and would be fine support `e and `u without the need for a RFC.  The risk as a breaking change is minimal and the value is apparent.  For line continuation, we suggest opening a separate issue to discuss that.

**Comment by DerpMcDerp** [comment]
I'm against the `` `u<hex>+`` without an end of hex sequence delimiter syntax because programming languages like that prone to mistakes when the escape code is followed by text that just happens to be a hex character. My preferred solution is something that resembles Perl's bracketed hex escapes like:

https://connect.microsoft.com/PowerShell/feedback/details/969678/add-character-literals-and-support-unicode-escape-characters-in-strings

Though I would modify the above proposal so that U+10000 to U+10FFFF produce `UnicodeCodePoint` objects instead of strings:

https://github.com/dotnet/corefxlab/blob/f89da2d8f363c9fbbe2786cdcbe0415a0a7a9ed9/src/System.Text.Primitives/System/Text/Encoding/UnicodeCodePoint.cs

**Comment by rkeithhill** [comment]
The way C# solves this, and what I was planning to implement, is that the token will always require four hex digits e.g.:
```csharp
Console.WriteLine("Testing \u0061\u0062\u0063def");
```
So in PowerShell that would be:
```powershell
"Testing `u0061`u0062`u0063def"
```
No worse (and yeah, no better) than C# in that regard.

RE UnicodeCodePoint, isn't that a bit of a corner case - using characters above BMP?  Not being sarcastic.  Just never run into a need for a Unicode surrogate pair in .NET or Windows.  And if you really need it, you can use:
```
PS> [System.Char]::ConvertFromUtf32(0x1F01C)
🀜
```

**Comment by DerpMcDerp** [comment]
"foo$([Char]::ConvertFromUtf32(0x1F01C))bar" doesn't work in places where constant strings are required

**Comment by rkeithhill** [comment]
Well, we could use `` `U<hex>{6}`` for this (or `` `U<hex>{8}`` just to be consistent with C#/C++).

**Comment by DerpMcDerp** [comment]
If the syntax is going to be `` `u`` Can you consider adding the `` `{(<hex>+<space>?)*}`` form also to support Perl style hex escapes:

``"`{61 62 63}"`` == "abc"

Since escapes often come in batches and it's annoying/cluttered to use `` `u0061`u0062`u0063 ``

It would be acceptable for me if the ranges outside the BMP could only be done inside the braced form so only the `` `u<hex>{4} `` and `` `{(<hex>+<space>?)*}`` form become valid.

**Comment by mklement0** [comment]
The ambiguity around the number of subsequent hex. digits could be resolved analogously to JavaScript (which, as an aside, also has _fixed_-length escapes), which supports `\u{...}` escapes, where the enclosing `{` and `}`  _unambiguously_ support  _between 1 and 8 hex. digits_:

E.g., with Node.js installed, try:  

```javascript
> node -pe "'Mot\u{f6}rhead'"
Motörhead
```

Translated to (hypothetical) PowerShell:

```powershell
> "Mot`u{f6}rhead"
Motörhead

> "Where's the `u{20ac}?"
Where's the €?
```

Given that `{...}` is already used to disambiguate identifiers in variable references, that seems like a natural fit.

**Comment by rkeithhill** [comment]
I'd prefer to stick closer to the C# approach.  PowerShell has been described by Jeffrey as having a "glideslope approach to C#".  I don't see **enough** value in the above to deviate from the C# approach.

**Comment by mklement0** [comment]
@rkeithhill: Funny you should say that, because just today [@lzybkr said](https://github.com/PowerShell/PowerShell/issues/3820#issuecomment-302750422) (yanked _somewhat_ out of context for dramatic effect): 

> PowerShell is not C# and it was not a goal to replicate all of C#'s capabilities.

With respect to PowerShell's _string literals and string interpolation_ staying close to C#, the HMS Pinafore of disparity has sailed a looooooong time ago.

**Comment by mklement0** [comment]
@rkeithhill:

Kidding aside:

* Personally, when I use PowerShell strings, I _never_ think of how strings work in C#, and from a _pure PowerShell perspective_, `` "`u{…}`" `` strikes me as a natural fit.

* What prompted Jason's quote was a context - class definitions - where a piece of "un-Powershell" C# syntax _did_ creep in (calling base-class constructors with `: base()`), prompting _mistaken expectations of more C#-like behavior_.

**Comment by rkeithhill** [comment]
Well I didn't mean to imply PowerShell and C# syntax were identical - just that there are areas where you can get some "transfer of learning".  That's the "glideslope" part that Jeffrey has referred to in the past.

So given one syntax that isn't demonstrably better than the other, I'd stick with the one that reflects the C# approach.  And the C# approach also reflects how other shells like Zsh/Bash do this e.g. `\u2620`.  This is useful when someone sees an example in say Bash, knows that PowerShell uses backtick instead of \, swaps out the \ for a backtick and gets it to work in PowerShell.  As Brad Abrams used to say back in the early days of .NET, as much as possible you want folks to fall into the "pit of success".

> Also note that this one flexible construct obviates the need for a separate "`e..." escape.

This is moot as support for the esc char has already been added to PowerShell.

I must have missed something with the class ctor chaining since `: base()` is supported in C# and in that regard, PowerShell is "similar" to C#.  However, unlike C#, PowerShell doesn't support `: this()` ctor chaining.

**Comment by mklement0** [comment]
> However, unlike C#, PowerShell doesn't support `: this()` ctor chaining.

That's exactly what the linked post is about. The point was that if you're familiar with C# syntax and see `: base()`, it's likely that you'll also expect `: this()` to work.

> This is moot as support for the esc char has already been added to PowerShell.

Thanks - I missed that, and I can definitely see that it's useful (I originally misinterpreted its purpose).

---

As for Bash: The **patchwork of escapes in Bash reflects the historical growth** of the code-point space:

```sh
printf '\x0a'        # byte value: up to 2 hex digits: the only kind supported up to Bash 3.x
printf '\u20ac'      # Unicode codepoint: up to 4 hex digits: Bash 4+; '€'
printf '\U00010300'  # Unicode codepoint: up to 8 hex digits: Bash 4+; '𐌀'
```
Note:  The above uses Bash's `printf` _builtin_ - not the external `/usr/bin/printf` utility.

**The only way to avoid ambiguity with the above is to use _exactly as many digits as supported by the respective escape_**, and the distinction between `\u` and `\U` is awkward.

Bash's [ANSI-C quoted string literals](https://www.gnu.org/software/bash/manual/bash.html#ANSI_002dC-Quoting), support `{...}` enclosure for `\x`, but, curiously, neither for `\u` or `\U`:

```sh
printf %s $'\x{a}'       # OK - \x supports  {...} enclosure 
printf %s  $'\u{20ac}'   # !! NOT supported
printf %s  $'\U{10300}'  # !! NOT supported 
```

The separate `\x` escape enables creation of arbitrary, single _bye_ values. With its ability to create byte values with the high bit set (e.g., `\xfc`), this enables creation of bytes that aren't valid in UTF-8, so at least hypothetically the `\x` escape serves a distinct purpose in Bash.

However, PowerShell, such a _byte_-value escape doesn't fit into the world of PowerShell's strictly-Unicode strings - and the need to create bytes can be handled with `[byte]` values, so the only thing that PowerShell _string_ interpolation must cover is _Unicode code-point escapes_ - **a _single_ concept that deserves a _single_ syntactic representation for simplicity:**

```powershell
"`u{a}"   # same as "`n"
"`u{fc}"   # 'ü'
"`u{20ac}"   # '€'
"`u{10300}"   # '𐌀' 
```

**Comment by mklement0** [comment]
<sup>Note:  
. This is a rewrite of a since-removed comment that lacked focus.  
. Also, I performed a flippantectomy on another still-extant comment  - my apologies, @rkeithhill -
 and merged it with a follow-up.</sup>

> However, unlike C#, PowerShell doesn't support `: this()` ctor chaining.

That's exactly what the linked post is about. The point was that if you're familiar with C# syntax and see `: base()`, it's likely that you'll also expect `: this()` to work.

> This is moot as support for the esc char has already been added to PowerShell.

Thanks - I missed that, and I can definitely see that it's useful (I originally misinterpreted its purpose).

---

As for transfer of learning: 

The `` `u `` \ `\u` correspondence is definitely helpful, but I think the _4-fixed-hex-digits_ demarcation is not worth adopting, or at least _not only_:

  * To support _all_ Unicode code points with the fixed-digit-count approach, the _uppercase_ variant `` `U ``  with _8_ fixed hex digits would have to be adopted as well, which is awkward in PowerShell, given its fundamentally _case-insensitive_ nature.  
(As an aside: the existing escapes such as `` `n` `` _are_ case-sensitive, though they arguably don't need to be.)

  * The patchwork of octal, `\x`, `\u`, `\U` escapes in various languages, including C#, exists for _historical_ reasons, and this is a chance to avoid carrying technical debt forward; the debt stems from:
      * gradual growth of the code-point space over time (ASCII (7-bit) > "ANSI" (8-bit) > UCS-2 (16-bit, BMP of Unicode only) > UTF-16 (all of Unicode, with surrogate pairs, 32-bit)
      * lack of support for embedded, variable-length string interpolation constructs.

 * A willingness to shed technical debt is already reflected in this thread's starting post considering _only_ `` `u ``, and not also `\x` or octal escapes.

  * The `{...}` form of delimiting is well established in PowerShell, can be found in several other scripting languages, and now even in C# 6's _interpolated strings_ - albeit not in the context of Unicode-character escapes.

  * To recap: Instead of requiring `` `u00fc `` (4 fixed digits), allow `` `u{fc}``, with between 1 and 8 digits to _unambiguously_ cover _all_ Unicode code points with a _single_ construct.

---

**A look at other languages:**

While `\x`, `\u`, and `\U` are common, there are differences with respect to:

 * whether a _fixed_ digit count is required or whether _up-to-a-max-count_ logic.
 * whether `{...}` is (also) supported for unambiguously enclosing a variable number of digits.

---

* C#, as discussed, has fixed-digit-count `\x`, `\u`, and `\U` only.

* POSIX-like shells:
   * The POSIX _standard_ mandates only _octal_ escapes, and then only in the context of specific utilities (e.g., for `<tab>`: `\011` in `printf`'s format string and in `echo`'s operands), not in _string literals_ generally.
  * `bash`, `ksh`, and `zsh` have a string-literal form `$'...'` that supports _variable-digit-count_, non-delimited `\x`, `\u`, and `\U` escapes (also implemented in the built-in versions of the `printf` and `echo` utilities).
  * `ksh`, commendably, additionally supports `{...}` for disambiguation in all 3 cases, whereas `bash`, curiously, only does so for the older `\x` escape. (`zsh` has no support at all).

* Ruby, Perl, and JavaScript (also) support `{...}`-delimited variable-digit-count escapes (though Perl's Unicode escapes look different).

**Comment by rkeithhill** [comment]
I have a PR for Unicode esc sequence coming.  Testing it right now.  This is the way I've implemented it so far:

* `` "`u2195"`` - simple, common use case which can be converted from C# or Bash by simply replacing `\` with `` ` ``.
* `` "`u{1F3F4}"`` - specify 1 to 8 hex digits.  A hex value `> 0xFFFF` and `<= 0x10FFFF` will generate a Unicode surrogate pair e.g.:
```posh
"Black flag: `u{1f3f4}" | Out-File .\unicode.txt -Encoding Unicode
```
The above generates the following file contents:
![image](https://cloud.githubusercontent.com/assets/5177512/26813694/1f77c1aa-4a3d-11e7-8181-29292d06e936.png)

One thing I'm trying to determine now is how widely should these esc sequences be supported.  For instance, double quoted strings and here strings are obvious but the following are perhaps not so obvious:

**Naming variables with Unicode chars**
```posh
PS> ${foo`u2195} = 'bar'
PS> gv foo*
Name                           Value
----                           -----
foo↕                           42
```
I'm inclined to say yes.  C# supports Unicode esc seqs in identifiers e.g.:
```c#
cl\u0061ss.st\u0061tic(true);
```
**And specifying arguments (generic tokens) with Unicode chars**
```posh
PS> Get-Command 7z`u2195
Get-Command : The term '7z↕' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ Get-Command 7z`u2195
+ ~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (7z↕:String) [Get-Command], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException,Microsoft.PowerShell.Commands.GetCommandCommand
```
Again I'm inclined to say yes to this as well.

Two other questions, I'd like to get input on.  It is worth having both forms (fixed length 4 digits and variable number of digits)?  And should the variable sequence form accept 8 hex chars or just 6?  The underlying API we use to convert the number to a Unicode surrogate pair only allows a max of 21-bit value (6 hex digits - although some 6 digit hex values will still be out-of-range).  However, I note that C#'s `` "\Uxxxxxxxx"`` supports 8 hex digits.  I'm wondering why they chose to support 8 digits?  Future proofing?

**Comment by rkeithhill** [comment]
I could use some feedback on these error messages.  Wording suggestions?

![image](https://cloud.githubusercontent.com/assets/5177512/26816131/69323338-4a4d-11e7-81d2-32a22573f632.png)

**Comment by Jaykul** [comment]
I've always assumed that they support only 4 or 8 simply because strings are 16bit unicode. 

I think you need to support them (at least) anywhere strings can be used (including if I use longer syntax, like `set-variable -Name ...` or `set-content -path function:....`) -- I mean, obviously yes to variable and function identifiers, and yes to parameters.... in fact, the argument example is just an implicit string, after all...

**Comment by rkeithhill** [comment]
Thanks Joel.  I'm getting close on this.  I've decided to support on 6 digits in the bracketed from.  This is from the C# docs:

> Unicode characters with code points above 0x10FFFF are not supported.

Just need to make sure my Pester tests pass.

**Comment by mklement0** [comment]
It's great that you're supporting the `` `u{...} `` syntax. 

> It is worth having both forms (fixed length 4 digits and variable number of digits)?

At the risk of repeating myself: I think this is an opportunity to ditch the fixed-digit form - which would also mean: fewer error messages to worry about :)

In that vein, re error messages: perhaps the following catch-all would then do (spread across 2 lines just for readability here):

    Invalid Unicode escape sequence:
    Must be `u{<hex-digits>}, using between 1 and 6 hex digits, up to a maximum of 10FFFF.

Re number of hex digits. in the `` `u{...} `` form:

6 digits are definitely enough in the foreseeable future (though the upcoming integration of the Martian alphabet may change that).
Among the languages / shells previously mentioned, Ruby does the same (all others support 8, presumably for the reason @Jaykul mentioned).

**Comment by DerpMcDerp** [comment]
@rkeithhill:

Does ``foo `u002Dbar`` or ``foo `u{2D}bar`` parse similar to `foo -bar` (i.e. gets treated as a potential keyword parameter rather than a string) ?

Does ``foo 2`u0030`` or ``2`u{30}`` parse similar `foo 20` (i.e. gets treated as number rather than a bareword/string) ?

Are you planning on supporting multi-escapes e.g. ``"`u{61 62 63}"`` means ``"`u{61}`u{62}`u{63}"`` ?

**Comment by mklement0** [comment]
@DerpMcDerp: Good questions, and without having looked at @rkeithhill's code and without wanting to speak for him, here's my $.02 with respect to how it _should_ work:

> Does foo `u002Dbar or foo `u{2D}bar parse similar to `foo -bar`

NO: Shell metacharacters such as `-` should only ever be recognized as _literals_.

> Does foo 2`u0030 or 2`u{30} parse similar foo 20

YES: In the parameter-value token-parsing stage, `` 2`u{30} `` should evaluate to `20` (`` 2`u0030 ``? Why, I never ...).

> gets treated as number rather than a bareword/string) ?

There's no such thing as a number _at that stage_. There's only a _token_ that will bind to a parameter, and that parameter's type determines the token's interpretation.

In other words:  `` 2`u{30} ``, a.k.a. `20` may turn into a number _or_ a string - it all depends on the type of the parameter it binds to.

**Comment by rkeithhill** [comment]
`` foo `u002dbar`` passes a string argument to the function (even if the function has a `$Bar` parameter defined.   So no.

``foo 2`u0030`` again passes a string argument to the function. So no.

I don't view either as problem.

Regarding support for multi-escapes, no that's what I've implemented with the current PR.

**Comment by mklement0** [comment]
@rkeithhill 

> foo 2`u0030 again passes a string argument to the function. So no.

Consider the following example:

```powershell
> function foo { param([int] $i) "[$i]" }; foo ('0' + 'x' + '10')
[16]
```

Clearly, what ended up binding to the `[int]` parameter was recognized as an _integer_, despite having been composed of string components, and having been recognized as a _hex_ representation.

While it seems ill-advised to represent token `20` as `` 2`u{30} ``, I see no reason why it should act differently.

Would your PR actually not work this way?

**Comment by rkeithhill** [comment]
The works as I've described.  It's like passing `` 7` 7`` or `` 7`t7`` as an argument.  That argument is parsed as a string due to it encountering an character esc seq.  The parser doesn't consider the "value" of the esc sequence.

**Comment by mklement0** [comment]
@rkeithhill: I don't think there's a real disagreement here, but there's terminology confusion: there's a token-parsing stage that results in a data type-agnostic sequence of characters that is then bound to a parameter, at which coercion to the target type happens.
Calling it a "string" at this stage is inviting confusion with  a `[string]`-typed parameter.

Just to clear things up:

 Will

```powershell
> function foo { param([int] $i) "[$i]" }; foo 20
[20]
```

NOT work the same as the following?

```powershell
> function foo { param([int] $i) "[$i]" }; foo 2`u{30}
[20]
```

**Comment by rkeithhill** [comment]
Yes but your second example coerces a string to an int. This is a more informative test:
```
PS> function foo($p) { "Type: $($p.GetType().FullName) value: $p" }
PS> foo 20
Type: System.Int32 value: 20
PS> foo 2`9
Type: System.String value: 29
PS> foo 2`u{30}
Type: System.String value: 20
```

**Comment by mklement0** [comment]
@rkeithhill: 

That is indeed an interesting illustration of what happens in the _absence_ of _typed_ parameters (note that _both_ my examples coerce to `[int]`).

So it would be more accurate to say: using `` ` `` in a token (such as to initiate a Unicode escape sequence) _defaults_ to `[string]` in the _absence of typing_ a parameter.

_With_  typing, no distinction is made - whether the token was a pure "number string" to begin with or resulted in one after expansion of escape sequences is immaterial.

**Comment by DerpMcDerp** [comment]
@keithhill:

Is ``$function`u003Apause`` equivalent to `$function:pause` or `$variable:function:pause` ?

**Comment by mklement0** [comment]
Ceterum censeo (I maintain my blissful ignorance of how thing are actually implemented): 

* Nothing that is the _result of an expansion (interpolation)_ should ever have _syntactical_ function. - metacharacters should only ever be recognized _as literals_.
(Treating arbitrary strings as commands is the domain of `Invoke-Expression`, which should be regarded with equal suspicion as `eval` in POISX-like shells.)

* Subexpressions (`$(...)`) should be _independent parsing contexts_ - whether or not they are embedded inside `"..."` (even though that is their primary purpose) should be immaterial.

**Comment by rkeithhill** [comment]
@DerpMcDerp - that doesn't parse.  You can create a function name with a Unicode esc seq:
```posh
function foo`u2195bar {}
foo↕bar  # invoke function
```
@mklement0 I mostly with you except for the last part. I don't consider it immaterial because the parameter parser is actually pretty smart about primitives e.g.:
```posh
PS> foo 20d
Type: System.Decimal value: 20
PS> foo 2.0
Type: System.Double value: 2
```
So the parser has determined the type of the argument before it is passed to the function.  Then the parameter binding because its process of type coercion (if necessary).  You can explore this with the `Trace-Command`.  Try this with your `foo` function passing in `10` and `'10'`.  Note this command seems to display more useful info in Windows PowerShell.  In PS Core, it's not so helpful.
```posh
PS> Trace-Command -Name ParameterBind*,TypeConversion -Expression {foo '10' } -PSHost
DEBUG: ParameterBinding Information: 0 : BIND arg [10] to parameter [i]
DEBUG: ParameterBinding Information: 0 :     Executing DATA GENERATION metadata:
[System.Management.Automation.ArgumentTypeConverterAttribute]
DEBUG: ParameterBinding Information: 0 :         result returned from DATA GENERATION: 10
DEBUG: ParameterBinding Information: 0 :     BIND arg [10] to param [i] SKIPPED
DEBUG: ParameterBinding Information: 0 : BIND arg [10] to parameter [i]
DEBUG: ParameterBinding Information: 0 :     Executing DATA GENERATION metadata:
[System.Management.Automation.ArgumentTypeConverterAttribute]
DEBUG: ParameterBinding Information: 0 :         result returned from DATA GENERATION: 10
DEBUG: ParameterBinding Information: 0 :     BIND arg [10] to param [i] SKIPPED
DEBUG: ParameterBinding Information: 0 : BIND arg [10] to parameter [i]
DEBUG: ParameterBinding Information: 0 :     Executing DATA GENERATION metadata:
[System.Management.Automation.ArgumentTypeConverterAttribute]
DEBUG: TypeConversion Information: 0 :         Converting "10" to "System.Int32".
DEBUG: TypeConversion Information: 0 :             Converting to integer.
DEBUG: ParameterBinding Information: 0 :         result returned from DATA GENERATION: 10
DEBUG: ParameterBinding Information: 0 :     COERCE arg to [System.Int32]
DEBUG: ParameterBinding Information: 0 :         Parameter and arg types the same, no coercion is needed.
DEBUG: ParameterBinding Information: 0 :     BIND arg [10] to param [i] SUCCESSFUL

PS> Trace-Command -Name ParameterBind*,TypeConversion -Expression {foo 10 } -PSHost
DEBUG: ParameterBinding Information: 0 : BIND arg [10] to parameter [i]
DEBUG: ParameterBinding Information: 0 :     Executing DATA GENERATION metadata:
[System.Management.Automation.ArgumentTypeConverterAttribute]
DEBUG: ParameterBinding Information: 0 :         result returned from DATA GENERATION: 10
DEBUG: ParameterBinding Information: 0 :     BIND arg [10] to param [i] SUCCESSFUL
```

**Comment by mklement0** [comment]
@rkeithhill 

Thanks, that was illuminating, and thanks for indulging my tangent; I've tried to summarize in more detail further below, which may be of interest to some.

* A token that contains a Unicode escape sequence will _default_ to type `[string]` (e.g., `` `u2{30} ``) - unlike a "straight number string" (e.g., `20`, which defaults to `[int]`).

* However, it is the _parameter_'s type that ultimately determines the token's type, _if specified_ (other than as `[object]`, `[psobject]`, `[pscustomobject]`). 

Therefore:

* With an _untyped_ parameter, the distinction matters: `20` becomes an `[int]`, and `` 2`u{30} `` becomes a `[string]`.

* With a _typed_ parameter, there is no _functional_ distinction between `20` and `` 2`u{30} ``, although if the target type is `[int]`,  the latter must undergo a second round of type conversion _behind the scenes_.

### Optional reading: Parameter parsing and types

Note: The following covers only tokens that are _literals_, except possibly with _quoting_ (which includes `` ` ``-based individual-character escapes and Unicode escape sequences) and _embedded_ variable references or subexpressions.

* The parameter parsers recognizes a few types _up front, in isolation, preliminarily_:

   * Something that is _unquoted_ (and has no embedded variable references/subexpressions) and can be parsed as a _number_ is converted to an appropriate number type (between `[int]` and `[double]`, including recognition of exponential notation, hexadecimal notation, and numbers with type suffixes `l` and `d`, but _not_ including negative numbers (tokens starting with `-`)).

   * Everything else becomes a `[string]`.

 * During parameter binding, that preliminary type _may_ be retained _as is_: 

   * If its type is the _same_ as the parameter's or it is a _derived_ type.

       * Note that the derived-type scenario doesn't apply here, given that the preliminary types are limited to number types (which are value types that can't be derived from) and `[string]`, which is a _sealed_ reference type.

   * If the parameter is _untyped_ or declared as `[object]` (this is a corollary of the above), but also if it is declared as `[psobject]` or - curiously - `[pscustomboject]`.

 * Otherwise, _coercion_ to the target type happens, which:
    * may be able to make use of the preliminary type (e.g., coercing a preliminary `[int]` to a `[long]` parameter),
     * may have to fall back to the raw string representation,
     * may fail altogether.

**Comment by rkeithhill** [comment]
Overall, pretty good summary.  I would tweak this sentence `unlike a "straight number string" (e.g., 20, which defaults to [int]).` to `unlike a character sequence corresponding to number (e.g., 20, which defaults to [int]).`  Saying it is a "number string" introduces some confusion - it is a number or a string.  :-)

> but not including negative numbers (tokens starting with -)

Careful on this. Remember that PowerShell parses in two modes: command and expression.  You've observed just the command mode.  You can change the behavior WRT `-<num>` by using a grouping expression or sub-expression e.g.:
```posh
PS> function z($n) { "`$n is $n, type is $($n.GetType().FullName)" }
PS> z -19
$n is -19, args[0] is , less than zero True
PS> z (-19)
$n is -19, type is System.Int32
```

> coercion to the target type happens, which:

The hoops that PowerShell jumps through to coerce one type to another are even more significant than that.  IIRC, PS will attempt to use Parse/TryParse methods if they exist as well as instances of ArgumentTransformationAttribute-derived classes (if the associated parameter is decorated with such attributes).  The latter allows users to inject custom logic into the type coercion process.  If the target type is a string, ToString() also fits into the process which is why if you pass a Process object to a parameter of type string, you'll get a value of `System.Diagnostics.Process (powershell)`.

I believe there may be other heuristics involved including using ctors with the correct signature and possibly other methods.

**Comment by mklement0** [comment]
Thanks, @rkeithhill, all good stuff; yes, I was only talking about argument-mode (which is what "about_Parsing" calls it, at least these days), and I've edited my post to make that clearer; I've also incorporated your other suggestions, except for the background on the advanced coercion techniques (I just point to your comment).

**Comment by rkeithhill** [comment]
Here's the latest version of parser error messages:
![image](https://user-images.githubusercontent.com/5177512/26999447-194ab036-4d57-11e7-83e1-f02255b9d7c1.png)

**Comment by mklement0** [comment]
These error messages are very clear and helpful.

If providing that level of specificity is the consensus, then I think they're great.

I wonder if the specificity is needed, however, given that someone even attempting to use this feature can be assumed to have a reasonable level of technical competence, which is why I suggested a single catch-all message earlier.

(But I'm probably venturing into bike-shedding territory here.)

**Comment by rkeithhill** [comment]
Thanks!  RE too specific, I remember a time when the PowerShell error messages were less than helpful.  Over time I think they've gotten better in terms of not only telling you what is wrong but also, when possible, suggesting a mitigation.  

Lately, I've been impressed with how well `Git` does this e.g.:
```
PS> git statuz
git: 'statuz' is not a git command. See 'git --help'.

Did you mean this?
        status
```
So while I agree we should keep the error messages as short as possible (otherwise TL;DR) I do think we should provide some help in terms of suggesting a fix.

**Comment by SteveL-MSFT** [comment]
@rkeithhill the PowerShell Team has discussed enhancing the current suggestion system to be more useful for missing or typos for cmdlets or parameters.  I didn't find an existing issue, so I opened https://github.com/PowerShell/PowerShell/issues/3988

**Comment by lzybkr** [comment]
The @PowerShell/powershell-committee reviewed this and we agreed to we do not want the fixed width syntax without braces - we only want the braced syntax.

Arguments against the fixed width syntax:
* It's harder to read
* It might be similar to C#, but it's not exactly the same, so cut and paste doesn't work anyway
* No possible ambiguity, e.g. `` `uabcd00 `` vs. `` `u{abcd}00``

**Comment by rkeithhill** [comment]
Works for me.  I'll try to get the updates finalized tomorrow night.  Just got home - spent all day of driving from Salt Lake City back to Colorado.

**Comment by rkeithhill** [comment]
I rebased and squashed down to a single commit the original support for both forms.  Then I pushed a commit with support for the fixed width form removed.

**Comment by DerpMcDerp** [comment]
@lzybkr What's the reason for the braced syntax being `` `u{abcd}`` instead of `` `{abcd}`` ?

**Comment by lzybkr** [comment]
@DerpMcDerp - the @PowerShell/powershell-committee did not discuss dropping the `u`, so I suppose it seemed obvious we needed something.

I see some advantages to keeping the `u` - it's reasonably descriptive, more similar to other languages, and it's a form of future proofing - it might be desirable to use that same syntax or very similar syntax for some future feature (I don't have anything specific in mind though.)

The only advantage I see to dropping the `u` - it's one less character, not much of an advantage.

**Comment by rkeithhill** [comment]
Both `` `e`` and `` `u{xxxxxx}`` have been [implemented](https://github.com/PowerShell/PowerShell/pull/3958), merged and documented.
